zerver/lib/actions.py

```diff
@@ -897,7 +897,9 @@ def do_send_messages(messages_maybe_none):
             presences    = presences)
         users = [{'id': user.id,
                   'flags': user_flags.get(user.id, []),
-                  'always_push_notify': user.enable_online_push_notifications}
+                  'always_push_notify': user.enable_online_push_notifications,
+                  'stream_push_notify': subscribed_to_stream_push_notifications(user, message['stream']),
```

Comment:
So the crux of the performance problem is this call; we're doing 1 database query per user subscribed to the stream, which is exactly what the test is meant to prevent us from doing, since it means if you send to a stream with 2000 subscribers, we're going to do 2000 database queries.

What you want to do instead is patch `get_recipient_user_profiles`, which is already querying all the relevant Subscription objects.  It should add to the `fields` list for streams the field `push_notifcations`, and then do the somewhat annoying logistics to pass the data back to the callers of `get_recipient_user_profiles`.  Maybe the right logistics is to return a dict `sub_dict` mapping {user_profile_id: sub_object}, and then the code here can access `sub_dict[user.id]['push_notifications']`.  It'll be a bit messier than that, because `sub_dict` will be empty for non-stream messages.

Can you try playing with that approach?  I'd definitely rebase onto master before doing more here because we just merged a PR that refactors `recipient_user_profiles` or its neighbors; can't remember exactly.